JBPM-6724: Stunner - Add BPMN Event Sub-Process

alepintus opened the pull request:
Hi @romartin , @wmedvede , @hasys ,
Event Sub-Process added and support for CSS property stroke-dasharray added.

Could you please take a look?

Thanks!

hasys commented:
Jenkins execute full downstream build

romartin commented:
ok to test

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-svg/kie-wb-common-stunner-svg-gen/src/main/java/org/kie/workbench/common/stunner/svg/gen/model/impl/StyleDefinitionImpl.java:
```diff
@@ -179,15 +194,30 @@ public Builder setFontSize(final double size) {
+        public Builder setStrokeDashArray(final String strokeDashArray) {
+            String[] tempStrokeDashArray = strokeDashArray.split(",");
+
+            int i = 0;
```
Same here @alepintus , this could be replaced by:
`Double[] doubles = Stream.of(tempStrokeDashArray)
                    .map(Double::valueOf)
                    .toArray(Double[]::new);`

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/EventSubprocess.java:
```diff
@@ -0,0 +1,180 @@
+public class EventSubprocess extends BaseSubprocess implements DataIOModel {
+
+    @Title
+    public static final transient String title = "Event Sub-Process";
```
Hey @alepintus , as from today hehe this is not mandatory, you can remove the `@Title` annotation and `title` class member as well - the literals go only in the i18n bundle.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/EventSubprocess.java:
```diff
@@ -0,0 +1,180 @@
+    public static final transient String title = "Event Sub-Process";
+
+    @Description
+    public static final transient String description = "An Event-Subprocess is placed within another Subprocess. It becomes active when its start event gets triggered and can interrupt the Subprocess context or run in parallel (non-interrupting).";
```
Same here, you can remove the `@Description` annotation and `description` class member as well - the literals go only in the i18n bundle.

alepintus commented:
Hi @romartin ,
code changed as suggested and missing test added.
Could you please take a look?

Thanks!

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

hasys commented:
Hi @alepintus, @romartin,

sorry, but it looks like some problems with networks on Jenkins side and it is not possible to build war file to test.

`15:06:21 Caused by: java.net.SocketTimeoutException: connect timed out`

CC @winklerm @mareknovotny

mareknovotny commented:
@hasys I don't I can reconnect slaves in kie CI openstack environment to be sure it is due this.

hasys commented:
Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/EventSubprocess.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Are there any tests for `Equals` and `hashCode` methods?

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-svg/kie-wb-common-stunner-svg-gen/src/main/java/org/kie/workbench/common/stunner/svg/gen/model/impl/StyleDefinitionImpl.java:
```diff
@@ -16,16 +16,21 @@
 
 public class StyleDefinitionImpl implements StyleDefinition {
 
+    // add css definition
```
It it TODO?

alepintus commented:
Hi @hasys,
thanks for comments! Added missing tests for Equals and hashCode methods for EventSubprocess and some other Events classes and deleted comment from StyleDefinitionImpl.

Could you please take a look?

Thanks!

hasys commented:
Index tests in other projects again. Can be ignored for now.

Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/StartMessageEvent.java:
```diff
@@ -125,8 +125,8 @@ public boolean equals(Object o) {
             return super.equals(other) &&
-                    executionSet.equals(other.executionSet) &&
-                    dataIOSet.equals(other.dataIOSet);
+                    (other.executionSet == null ? executionSet == null : other.executionSet.equals(executionSet)) &&
```
Sorry, just found a "better way":
```java
Objects.equals(executionSet, other.executionSet);
```
but it's optional.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/BaseThrowingIntermediateEvent.java:
```diff
@@ -178,11 +178,11 @@ public int hashCode() {
+                    (other.dataIOSet == null ? dataIOSet == null : other.dataIOSet.equals(dataIOSet)) &&
+                    (other.backgroundSet == null ? backgroundSet == null : other.backgroundSet.equals(backgroundSet)) &&
+                    (other.fontSet == null ? fontSet == null : other.fontSet.equals(fontSet)) &&
+                    (other.dimensionsSet == null ? dimensionsSet == null : other.dimensionsSet.equals(dimensionsSet));
```
It looks like `protected final Set<String> labels = new HashSet<String>();` is not part of equals/hashCode methods, same as in BaseSubprocess.java. Not sure is it correct. @romartin any ideas?

romartin replied:
yes well probably it's a bit confusing even for me now the labels. I mean actually the labels are part of the bean's state, so makes sense checking the equality too, on the other hand, the labels collection is mutable but we can ensure (for now) that labels in Stunner are not being changed from the default ones for each node. So adding the labels into the equals  makes sense but will have no effect at this point,  because for example 2 UserTask's will have always same labels...
This is related with the job that @evacci and @hasys are doing on another PR, so proably you guys can take the right decision...

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/BaseThrowingIntermediateEvent.java:
```diff
@@ -178,11 +178,11 @@ public int hashCode() {
+                    (other.fontSet == null ? fontSet == null : other.fontSet.equals(fontSet)) &&
+                    (other.dimensionsSet == null ? dimensionsSet == null : other.dimensionsSet.equals(dimensionsSet));
         }
         return false;
```
It looks like equals tests don't test condition NOT (o instanceof BaseThrowingIntermediateEvent).

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -126,7 +137,134 @@ public void testEndTerminateEventHashCode() {
+
+        assertEquals(a, a);
+        assertEquals(a, b);
+        assertFalse(a.equals(c));
```
In my opinion: `assertNotEquals(a, c)` is more verbose (here and for all further assertFalse)

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -126,7 +137,134 @@ public void testEndTerminateEventHashCode() {
+        EndSignalEvent a = builder.build();
+        builder = new EndSignalEvent.EndSignalEventBuilder();
+        EndSignalEvent b = builder.build();
+        assertEquals(a.hashCode(), b.hashCode());
```
I guess, we can merge this test with Equals test and add checks for all possible cases.
I mean like:
```java
assertEquals(a, a);
assertEquals(a.hashCode(), a.hashCode());
assertEquals(a, b);
assertEquals(a.hashCode(), b.hashCode());
assertNotEquals(a, c);
assertNotEquals(a.hashCode(), c.hashCode());
```
and so on.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -126,7 +137,134 @@ public void testEndTerminateEventHashCode() {
+        EndMessageEvent a = builder.build();
+        builder = new EndMessageEvent.EndMessageEventBuilder();
+        EndMessageEvent b = builder.build();
+        assertEquals(a.hashCode(), b.hashCode());
```
The same as about (assertNotEquals/merge hash and equals tests)

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -162,13 +300,138 @@ public void testIntermediateTimerEventEquals() {
+        IntermediateMessageEventCatching a = builder.build();
+        builder = new IntermediateMessageEventCatching.IntermediateMessageEventCatchingBuilder();
+        IntermediateMessageEventCatching b = builder.build();
+        assertEquals(a.hashCode(), b.hashCode());
```
The same as about (assertNotEquals/merge hash and equals tests)

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -162,13 +300,138 @@ public void testIntermediateTimerEventEquals() {
         builder = new IntermediateTimerEvent.IntermediateTimerEventBuilder();
         IntermediateTimerEvent b = builder.build();
-        assertTrue(a.hashCode() == b.hashCode());
+        assertEquals(a.hashCode(), b.hashCode());
```
The same as about (assertNotEquals/merge hash and equals tests)

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -336,7 +623,71 @@ public void testStartSignalEventHashCode() {
+        StartMessageEvent a = builder.build();
+        builder = new StartMessageEvent.StartMessageEventBuilder();
+        StartMessageEvent b = builder.build();
+        assertEquals(a.hashCode(), b.hashCode());
```
The same as about (assertNotEquals/merge hash and equals tests)

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -252,9 +515,33 @@ public void testReusableSubprocessHashCode() {
+        EventSubprocess a = builder.build();
+        builder = new EventSubprocess.EventSubprocessBuilder();
+        EventSubprocess b = builder.build();
+        assertEquals(a.hashCode(), b.hashCode());
```
I guess, test for EventSubprocess equals/hashCode should be quite similar to Events tests.

romartin commented:
hey @alepintus , yeah good job. Well some of the @hasys comments make sense although are minor, so please lemme know if you're doing to provide a new commit or can I merge it as it is. Thx

alepintus commented:
Hi @romartin @hasys ,
tests changed as suggested and added labels check on equals methods.

Could you please take a look?

Thanks!

alepintus commented:
Added labels on hashCode methods as well.

romartin commented:
Jenkins execute full downstream build

romartin commented:
Great thanks @alepintus , tested again although changes were not related to the UI, all good! :+1:
